### ISSUE TITLE
Regarding Report Portal Resilience 

### ISSUE THREAD

**Issue by priyankachaturvedi30** [reportportal/reportportal#1365]
@DzmitryHumianiuk : Currently, we are using Docker for running our Report Portal instance.  We have ~30 projects and executions on all of these are running in parallel. There are instances where we have observed that the RP server goes down and our run gets interrupted. 
We want to implement a load balancer to avoid such issues - something similar to what is described in https://github.com/reportportal/reportportal/issues/693
Out of the following services, which one should be replicated across servers? 
![image](https://user-images.githubusercontent.com/50797731/114727292-b6d61400-9d5b-11eb-8765-75d30b3d2b6f.png)

Do we have any documentation to implement the same? Any recommended best practices one should follow while doing so?

**Comment by priyankachaturvedi30** [comment]
@evjlobanova : Can you help me with the above query ?

**Comment by DzmitryHumianiuk** [comment]
@priyankachaturvedi30 https://github.com/reportportal/kubernetes should describe it.
You need to replicate API services.

@irynakozak2 @Cyberglamdring  could you please point to the right documentation?

**Comment by priyankachaturvedi30** [comment]
@irynakozak2  @Cyberglamdring : Can you help me with the documentation that @DzmitryHumianiuk  is suggesting so that I can move ahead with this approach ?

**Comment by Cyberglamdring** [comment]
Hello @priyankachaturvedi30,

You can install Report Portal on Kubernetes. At now we support Kubernetes version **1.16** – **1.19**. As a Load Balancer we use NGINX ingress. Also, we recommend using [2 replicas](https://github.com/reportportal/kubernetes/blob/master/reportportal/values.yaml#L63) of API services on different nodes for great fault tolerance. This [documentation guides](https://github.com/reportportal/kubernetes/tree/master/reportportal) you to install ReportPortal step by step.

### 🏛️ Architecture

There are two options:

- **3 nodes**: 1st node for RabbitMQ service, 2nd and 3rd nodes for API. The database can use different VM or Cloud service (For example Amazon RDS). To ensure fault tolerance and reduce the load on the application use the labels:
```
kubectl label nodes <NODE-1> service=api
kubectl label nodes <NODE-2> service=api
kubectl label nodes <NODE-3> service=rabbitmq
```

- **4 nodes**: 1st node for RabbitMQ service, 2nd and 3rd nodes for API. 4th node for Database service. 
```
kubectl label nodes <NODE-1> service=api
kubectl label nodes <NODE-2> service=api
kubectl label nodes <NODE-3> service=rabbitmq
kubectl label nodes <NODE-4> service=db
```

Recommended [hardware options](https://reportportal.io/docs/Optimal-Performance-Hardware) (CPU/Mem).

**Comment by priyankachaturvedi30** [comment]
So I was having a look at the architecture you shared above - https://github.com/reportportal/reportportal/blob/master/public/rp_repo_structure.png
Over here I can see -
```
Gateway and Load Balancer
- Traefik (Docker Config)
- NGNIX (Kubernetes Config)
```

I have deployed Report Portal with Docker setup. Would I be required to deploy it using Kubernetes (as specified above), or Traefik can do the load balancing for me ?

**Comment by priyankachaturvedi30** [comment]
@Cyberglamdring @DzmitryHumianiuk : Can you help me with my last comment so that I can move a step ahead in this direction

**Comment by Cyberglamdring** [comment]
Hello @priyankachaturvedi30,

Currently, Traefik load balancing the only HTTP requests. One way to use Docker and Traefik is to upscale your VM. Adding a second API service to one VM does not make sense, since both services will failure when resources are exhausted. 

We recommend using the Kubernetes + NGINX.

**Comment by priyankachaturvedi30** [comment]
@Cyberglamdring : Can we plan to use a second VM to replicate the API service? Would that work in case we want to have a load balancer in the Docker setup? Can you also help us with steps to replicate Api service on same/different VM, so that we can explore further.

**Comment by priyankachaturvedi30** [comment]
@Cyberglamdring : Thanks for the details. it was really helpful. I tried setup up my 2 vms in the fashion you described above. However I face the following challenges : 
1.) On VM2 -> my Api service says gives the following error related to keystore/migrations - 
![image](https://user-images.githubusercontent.com/50797731/116708073-f9028500-a9ec-11eb-9c02-8b8ceaad6eee.png)
2.) On VM1 -> Two of my services api and authorization


 keeps on restarting possibly because of connection issue (which i am debugging)

However, can you please look at the issue I am facing on VM2. 
Also, I am attaching docker-compose file used on both the VM's for your reference to double-check if I am missing something around here.
[docker-compose_VM1.txt](https://github.com/reportportal/reportportal/files/6406060/docker-compose_VM1.txt)
[docker-compose_VM2.txt](https://github.com/reportportal/reportportal/files/6406061/docker-compose_VM2.txt)

**Comment by priyankachaturvedi30** [comment]
@Cyberglamdring : Did you get a chance to check the above ? Also, I was wondering that we started UI service on first vm, considering the initial use case about resilience, if first vm goes down (which has the ui service running i.e the IP of RP instance which we need to push the log to, how would it basically be able to redirect them on second vm? Am i missing something here ?

**Comment by Cyberglamdring** [comment]
@priyankachaturvedi30, error `keystore/migrations` is doesn't matter. It will be fixed in a future release. Try running reporting and make sure everything is working fine.

About resilience. All important requests pass between API and Database.

**Comment by priyankachaturvedi30** [comment]
I understand @Cyberglamdring . But consider a scenario -
1. I have an  Automation Server (Client) and 2 RP instances (VM1 and VM2).
2. Now my Automation Server has to be mapped to VM1 IP (rp.endpoint) in order to send logs to it.
3. We do have done the setup where we have two VM's ready with api services replicated on them.
But I still do not understand that for my Step - 2, if VM1 is down (i.e. rp.endpoint is not accessible only) , then how would Step - 3 be able to help us ?

**Comment by priyankachaturvedi30** [comment]
I was trying to deep dive into the load balancer config and understood it could be done through traefik. Is there some additonal configuration that i need to do in order to achieve that? Following my previous comment...How can i ensure that if VM1 is down then my requests are re-directed to VM2 . Help much appreciated ! @Cyberglamdring

**Comment by DzmitryHumianiuk** [comment]
Closing this issue due to inactivity. 

If this still remains a concern or if there are new developments, please feel free to reopen or create a new issue. 

Thank you for your understanding and contributions to the project!
